Wipeout 3.10: Add new field to rights models

vojtechjelinek opened the pull request:
## Explanation
This PR adds two verification jobs `AddAllUserIdsVerificationJob` and `AddAllUserIdsSnapshotsVerificationJob`. These are used to create and fill the `[Collection|Exploration|Topic]RightsAllUsersModel` so that they can be verified by the `GaeIdNotInModelsVerificationJob`. The `[Collection|Exploration|Topic]RightsAllUsersModel` are filled by the user IDs from `[Collection|Exploration|Topic]RightsModel` and `[Collection|Exploration|Topic]RightsSnapshotContentModel`.

The `AddAllUserIdsVerificationJob` also verifies if the rights model contains all the user IDs that were in the snapshots, (we don't expect the snapshots to contain any user ID that is not in the parent rights model since there is no way to remove user ID from the `[Collection|Exploration|Topic]RightsModel` that means that all the user IDs should already be present there).

The need for these models is based on the fact that the `[Collection|Exploration|Topic]RightsSnapshotContentModel` are encapsuled thus their user ID fields cannot be indexed.

Also, removed the old `TranslatorToVoiceArtistOneOffJob`.

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

vojtechjelinek committed:
Add all_user_ids field

vojtechjelinek committed:
Prepare migration fields

vojtechjelinek committed:
Add job for filling all_user_ids

vojtechjelinek committed:
Add comments

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

# Conflicts:
#	core/domain/activity_jobs_one_off.py

vojtechjelinek committed:
Add rights snapshots migration jobs

vojtechjelinek committed:
Fix backend tests

vojtechjelinek commented:
@seanlip PTAL, there might be some backend tests failing but these should be just simple fixes.

seanlip commented on core/domain/activity_jobs_one_off.py:
```diff
@@ -59,6 +52,145 @@ def reduce(key, values):
 
 
+class AddAllUserIdsOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """For every right model merge the data from all the user id fields
```
right --> rights

vojtechjelinek replied:
Done.

seanlip commented on core/domain/activity_jobs_one_off.py:
```diff
@@ -59,6 +52,145 @@ def reduce(key, values):
 
+class AddAllUserIdsOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """For every right model merge the data from all the user id fields
+    together and put them in new all_user_ids field.
```
new --> a new

vojtechjelinek replied:
Done.

seanlip commented on core/domain/activity_jobs_one_off.py:
```diff
@@ -59,6 +52,145 @@ def reduce(key, values):
+        commit_message = 'Add new all_user_ids field'
+        commit_cmds = [{'cmd': 'add_new_field', 'field_name': 'all_user_ids'}]
+
+        unifed_rights_classes = (
```
What is "unifed"? I don't think that's a word.

Also, maybe consider inlining this, I'm not sure you need a separate variable for it given the context.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/activity_jobs_one_off.py:
```diff
@@ -59,6 +52,145 @@ def reduce(key, values):
+
+
+class AddAllUserIdsSnapshotsOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """For every snapshot of right model merge the data from all the user id
```
right model --> a rights model,

 (note the comma at the end)

vojtechjelinek replied:
Done.

seanlip commented on core/domain/activity_jobs_one_off.py:
```diff
@@ -59,6 +52,145 @@ def reduce(key, values):
+    def map(rights_model):
+        """Implements the map function for this job."""
+        commit_message = 'Add new all_user_ids field'
+        commit_cmds = [{'cmd': 'add_new_field', 'field_name': 'all_user_ids'}]
```
If this is a structural change, I don't think this should be made as a commit (which will end up creating new historical models). Shouldn't we be surgically updating all the existing versions of all the models in the datastore instead (but not create new ones)?

vojtechjelinek replied:
Done.

seanlip commented on core/domain/activity_jobs_one_off_test.py:
```diff
@@ -108,6 +110,317 @@ def mock_add_documents_to_index(docs, index):
+            [self.USER_1_ID, self.USER_2_ID],
+            topic_models.TopicRightsModel.get_by_id(self.TOP_1_ID).all_user_ids)
+
+    def test_multiple(self):
```
Make the test name more explicit.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/activity_jobs_one_off_test.py:
```diff
@@ -108,6 +110,317 @@ def mock_add_documents_to_index(docs, index):
+            [self.USER_1_ID, self.USER_2_ID, self.USER_3_ID],
+            topic_models.TopicRightsModel.get_by_id(self.TOP_1_ID).all_user_ids)
+
+    def test_multiple(self):
```
Make the test name more explicit.

vojtechjelinek replied:
Done.

seanlip reviewed (changes requested):
Thanks @vojtechjelinek. I have some concerns about this:

- If we're only updating the latest model, wouldn't this cause a mismatch between fields in old versions and newer snapshots of rights models? Or do we not persist all_user_ids in the historical snapshots?
- Please add tests specifically for flows like "X is added to exploration, X is subsequently removed from exploration, all_user_ids still includes X".
- Could you explain again why we are doing this, just so the context is self-contained within this PR?

Thanks!

vojtechjelinek committed:
Address comments

vojtechjelinek committed:
Fix backend tests

vojtechjelinek committed:
Rework

vojtechjelinek committed:
Add validation jobs and fix tests

vojtechjelinek reviewed (commented):
@seanlip PTAL, I need to add some backend tests to improve the coverage but the code is conceptualy complete.

vojtechjelinek committed:
Fix coverage

seanlip commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -3368,6 +3390,14 @@ def _get_external_id_relationships(cls, item):
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return {'topic_rights_ids': (topic_models.TopicRightsModel, [item.id])}
```
Break after {, and rearrange as in the examples above.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -516,3 +516,175 @@ class ModelsUserIdsHaveUserSettingsExplorationsVerificationJob(
+        collection rights model.
+        """
+        content_dict = (
+            collection_models.CollectionRightsModel.transform_dict_to_valid(
```
Can we rename transform_dict_to_valid to convert_to_valid_dict? I think that's more semantically correct (the ordering of the words is better and I think "convert" works better than "transform" here since it feels less arbitrary -- i.e. we're converting to a specific format rather than making an arbitrary transformation).

vojtechjelinek replied:
Done.

seanlip commented on core/storage/collection/gae_models.py:
```diff
@@ -434,6 +433,63 @@ def export_data(cls, user_id):
         }
 
 
+class CollectionRightsAllUsersModel(base_models.BaseModel):
```
If this and other models are meant to have a short lifetime of existence, please say so up here in the top-level docstring and write a todo/issue for deleting them. The issue should explain clearly the deprecation/deletion procedure.

Also, how are we planning to keep these models up to date as new users get rights? I don't see any code for doing that.

vojtechjelinek replied:
Done.

We don't since these will only be used for the user ID migration.

seanlip replied:
Say the migration fails (for some reason) and we have to bring the server up again. What happens to these models?

Ditto elsewhere.

vojtechjelinek replied:
They stay there and we can reupdate them again with the `AddAllUserIdsVerificationJob` and `AddAllUserIdsSnapshotsVerificationJob`.

seanlip commented on core/storage/collection/gae_models.py:
```diff
@@ -434,6 +433,63 @@ def export_data(cls, user_id):
+    def migrate_model(cls, unused_old_user_id, unused_new_user_id):
+        """This model is used to verify the user ID migration so it will be
+        filled by the AddAllUserIdsOneOffJob and AddAllUserIdsSnapshotsOneOffJob
+        after the migration is done, thus it shouldn't be migrated here.
```
I had some trouble understanding this docstring. Can you explain what "here" means in "shouldn't be migrated here"? Also explain in more detail why the migration is unnecessary.

Also -- I am not sure I understand why we even need this and the above method if the model has deletion policy DELETE?

vojtechjelinek replied:
Done. This method is needed because `USER_ID_MIGRATION_POLICY = CUSTOM` (and that is because we need the `verify_model_user_ids_exist` to be run.

seanlip commented on core/storage/exploration/gae_models.py:
```diff
@@ -477,6 +477,63 @@ def export_data(cls, user_id):
         }
 
 
+class ExplorationRightsAllUsersModel(base_models.BaseModel):
```
Same notes as above.

vojtechjelinek replied:
Done.

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -405,6 +405,10 @@ class TopicRightsModel(base_models.VersionedModel):
     # The user_ids of the managers of this topic.
     manager_ids = ndb.StringProperty(indexed=True, repeated=True)
+
+    # The user_ids of users who are (or were in history) members of manager_ids.
```
Why is this still here? Please make sure to proofread the PR diff carefully before requesting review.

vojtechjelinek replied:
Removed, I have proofread this PR but missed this :(

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -531,3 +538,60 @@ def _trusted_commit(
         ).put()
+
+
+class TopicRightsAllUsersModel(base_models.BaseModel):
```
Same comments as above.

vojtechjelinek replied:
Done.

seanlip reviewed (changes requested):
Thanks @vojtechjelinek, took a pass.

One question -- have you updated the wipeout design doc with this info? If not please do so. Maybe do it as a suggestion and loop me in for review.

Also, please document the purpose etc. of this PR more clearly in the opening comment. Provide context on the new models and their lifetimes, and the procedure surrounding them. This is important for giving reviewers context (we talked about this in our last meeting). Currently the documentation/explanation around this PR is a bit too skimpy.

Thanks!

seanlip commented:
Please also make sure to reply to all comments here: https://github.com/oppia/oppia/pull/8505#pullrequestreview-348486363

vojtechjelinek committed:
Address comments

vojtechjelinek reviewed (commented):
> If we're only updating the latest model, wouldn't this cause a mismatch between fields in old versions and newer snapshots of rights models? Or do we not persist all_user_ids in the historical snapshots?

We are now updating totally separate models.

> Please add tests specifically for flows like "X is added to exploration, X is subsequently removed from exploration, all_user_ids still includes X".

These are present, for example, https://github.com/oppia/oppia/pull/8505/files#diff-7374eba96e576667133cfb813577a0b7R1375

> Could you explain again why we are doing this, just so the context is self-contained within this PR?

Added explanation to the top.

> One question -- have you updated the wipeout design doc with this info? If not please do so. Maybe do it as a suggestion and loop me in for review.

Since now this only relates to the user ID migration I will add it to the User ID migration doc.

vojtechjelinek committed:
Address comments

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Rename tests

vojtechjelinek committed:
Fix tests

vojtechjelinek commented:
Some refactor is required for this PR, will do that tomorrow.

vojtechjelinek committed:
Address comments

vojtechjelinek committed:
Refactor jobs

vojtechjelinek commented:
I have moved the verification of the fact that the parent rights model contains all the user IDs that are in the snapshot models into the `AddAllUserIdsVerificationJob` because when the verification was in `AddAllUserIdsSnaphotsVerificationJob` we couldn't be sure what is the order of the models being verified and that produced errors.

vojtechjelinek commented:
@seanlip PTAL

vojtechjelinek committed:
Fix backend tests

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -516,3 +516,269 @@ class ModelsUserIdsHaveUserSettingsExplorationsVerificationJob(
+    def _add_collection_user_ids(rights_model, all_users_model):
+        """Compare the existing CollectionRightsAllUsersModel with the user IDs
+        in the CollectionRightsModel, if some of the user IDs from snapshots are
+        not in the parent rights model return them.
```
Seems like you do more than that -- you change the datastore too? It's important to explain this, because at first I assumed this was just a GET method...

Shouldn't we report an error and then look into it?

Ditto below.

vojtechjelinek replied:
Done. We return them and then set `SUCCESS-NOT_SUBSET` as the result so we can analyze this after the migration.

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -456,6 +447,8 @@ def migrate_model(cls, old_user_id, new_user_id):
                 new_user_id if manager_id == old_user_id else manager_id
                 for manager_id in model.manager_ids]
+            # These will be set by the AddAllUserIdsVerificationJob.
+            model.all_user_ids = []
```
This is wrong, right? We should not be modifying the TopicModel...

vojtechjelinek replied:
Done.

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -484,6 +477,14 @@ def verify_model_user_ids_exist(self):
             user_ids, include_deleted=True)
         return all(model is not None for model in user_settings_models)
 
+    def compute_snapshot(self):
```
Why do we need this for this model and not the other two?

Also please compare the TopicModel code to the others in general -- I would expect the changes to the three models to be broadly similar.

vojtechjelinek replied:
Done.

seanlip reviewed (changes requested):
Thanks @vojtechjelinek! It's pretty close IMO, I took a pass.

vojtechjelinek commented on core/storage/topic/gae_models.py:
```diff
@@ -424,19 +425,9 @@ def has_reference_to_user_id(cls, user_id):
-                cls.SNAPSHOT_METADATA_CLASS.exists_for_user_id(user_id))
+        return (cls.query(
+            cls.manager_ids == user_id).get(keys_only=True) is not None
+                or cls.SNAPSHOT_METADATA_CLASS.exists_for_user_id(user_id))
```
These changes were already done in the exp and col models before, that is why the changes differ a little bit for the topic model.

vojtechjelinek committed:
Address comments

seanlip reviewed (commented):
Thanks, no further code concerns. Do we need to run this job on the test server?

vojtechjelinek commented:
@seanlip It will be part of the migration, I will send you the instructions when these should be run, firstly we need to run `ReplaceAdminIdOneOffJob` and `GeneralFeedbackThreadUserOneOffJob` on the backup server.

vojtechjelinek commented:
@ankita240796 @aks681 @nithusha21 PTAL

ankita240796 commented on core/storage/collection/gae_models.py:
```diff
@@ -434,6 +433,69 @@ def export_data(cls, user_id):
+    """
+    # The user_ids of users who are (or were in history) members of owner_ids,
+    # editor_ids, voice_artist_ids or viewer_ids in corresponding rights model.
+    all_user_ids = ndb.StringProperty(indexed=True, repeated=True)
```
We should have a validation function in prod_validation to ensure that these ids correspond to valid UserSettingsModel.

vojtechjelinek replied:
Done.

ankita240796 commented on core/storage/topic/gae_models.py:
```diff
@@ -531,3 +522,65 @@ def _trusted_commit(
+    """
+    # The user_ids of users who are (or were in history) members of manager_ids
+    # in corresponding rights model.
+    all_user_ids = ndb.StringProperty(indexed=True, repeated=True)
```
Ditto.

ankita240796 commented on core/storage/exploration/gae_models.py:
```diff
@@ -477,6 +477,69 @@ def export_data(cls, user_id):
+    """
+    # The user_ids of users who are (or were in history) members of owner_ids,
+    # editor_ids, voice_artist_ids or viewer_ids in corresponding rights model.
+    all_user_ids = ndb.StringProperty(indexed=True, repeated=True)
```
Ditto.

vojtechjelinek replied:
Done.

ankita240796 commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -1098,6 +1098,17 @@ def _get_external_id_relationships(cls, item):
+    def _get_external_id_relationships(cls, item):
+        return {
+            'collection_rights_ids': (
+                collection_models.CollectionRightsModel, [item.id])
```
Add a mapping in `CollectionModelValidator` as well i.e. for each `CollectionModel`, we should have a `CollectionRightsAllUsersModel`.

vojtechjelinek replied:
Done.

ankita240796 commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -1884,6 +1895,17 @@ def _get_external_id_relationships(cls, item):
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return {
+            'exploration_rights_ids': (
```
Ditto.

vojtechjelinek replied:
Done.

ankita240796 commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -3368,6 +3390,17 @@ def _get_external_id_relationships(cls, item):
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return {
+            'topic_rights_ids': (
```
Ditto.

vojtechjelinek replied:
Done.

ankita240796 reviewed (changes requested):
Hi @vojtechjelinek, Thanks for the PR, I have added a few comments, PTAL.

nithusha21 commented on core/domain/activity_jobs_one_off.py:
```diff
@@ -24,16 +24,9 @@
-            models.NAMES.exploration, models.NAMES.question,
-            models.NAMES.skill, models.NAMES.story,
-            models.NAMES.topic]))
+(collection_models, exp_models, topic_models) = (
```
Shouldn't this have been flagged by the linter as variable not used or something like that?

seanlip replied:
I *think* it will detect regular imports, but not this sort of import. It would be nice to add it as a check though. I've added it to the list in #8423.

nithusha21 reviewed (approved):
Took a look as a codeowner. Nothing much to add, just a query!

vojtechjelinek committed:
Add validations

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

# Conflicts:
#	core/storage/collection/gae_models.py
#	core/storage/exploration/gae_models.py
#	core/storage/topic/gae_models.py
#	core/storage/topic/gae_models_test.py

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Merge remote-tracking branch 'remotes/upstream/develop' into wipeout-all-users

vojtechjelinek commented:
@ankita240796 @seanlip PTAL

aks681 reviewed (approved):
lgtm as codeowner.

ankita240796 commented on core/domain/prod_validation_jobs_one_off_test.py:
```diff
@@ -1704,6 +1707,104 @@ def test_invalid_collection_version_in_model_id(self):
+            update_datastore_types_for_mock_datetime()
+            run_job_and_check_output(self, expected_output, sort=True)
+
+    def test_missing_exploration_model_failure(self):
```
Please update the tests as well to include checks for missing user models and missing user rights models.

vojtechjelinek replied:
Done.

ankita240796 reviewed (changes requested):
PTAL @vojtechjelinek, Thanks!

vojtechjelinek committed:
Add tests

ankita240796 reviewed (approved):
LGTM, Thanks @vojtechjelinek!

vojtechjelinek committed:
Add jobs to job registry

vojtechjelinek committed:
Add support for deprecated gae_user_id

vojtechjelinek commented:
@seanlip PTAL at the latest commit, the issue with `gae_user_id` should be fixed there.

seanlip commented:
Commit looks OK to me. But as mentioned in the email thread, might want to investigate the discrepancy in what we expected from the original job vs the results we saw...

vojtechjelinek committed:
Limit number of models in transaction

vojtechjelinek committed:
Remove unused imports

vojtechjelinek committed:
Use python utils

vojtechjelinek commented:
@seanlip PTAL at last three commits, I'm fixing the transaction problem there.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -40,6 +41,8 @@
     exp_models.ExplorationCommitLogEntryModel,
     exp_models.ExplorationSnapshotMetadataModel]
 
+NUMBER_OF_MODELS_IN_TRANSACTION = 20
```
Add MAX_ before this constant.

Also move the constant to within _copy_model_with_new_id_and_user_id() -- we don't want others to use this or for it to be a general thing.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -103,12 +106,18 @@ def _copy_model_with_new_id_and_user_id(
 
-        def _replace_models():
+        def _replace_models(new_models_sub, old_models_sub):
             """Replace old models with new ones."""
```
Missing args docstring.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -103,12 +106,18 @@ def _copy_model_with_new_id_and_user_id(
+                new_models_sub, update_last_updated_time=False)
+            model_class.delete_multi(old_models_sub)
+
+        for i in python_utils.RANGE(
```
Add an explanation for why we limit the number of models.

Also, this introduces a possible failure case. What happens if the last model to be migrated fails somehow -- what are our remediation steps in that case?

Also, how many models does this for loop currently operate over?

vojtechjelinek replied:
Done.

If anything fails we can always repeat the migration.

I'm not sure, the models that are in this category are for example models that track the progress of the user in an exploration, meaning that there could be around 100 for one user. But to get more precise results we would have to query the datastore.

seanlip replied:
> Also, this introduces a possible failure case. What happens if the last model to be migrated fails somehow -- what are our remediation steps in that case?

Does repeating the migration actually work? If we are partially in a migration of user IDs then in the second run of map(), wouldn't " if user_model.id != user_model.gae_id:" have the possibility of already being True, leading to some orphaned models?

vojtechjelinek replied:
Hmmm, true this can happen, but this will be checked by the verification jobs that will be run after the migration. So it is true that the silent errors can happen but will be detected by the verification job and most of the errors we had were visible in the logs.

seanlip replied:
Sure, I understand that. But suppose, when we run this job in production, that the above scenario happens. What should we do then? We should think this through as part of our procedure.

vojtechjelinek replied:
The verification will fail, and we restore the backed up data.

vojtechjelinek replied:
The verification will fail, and we restore the backed up data.

seanlip replied:
I think that's a problem. Restoring data takes quite a while and this may lead to extended server downtime. Typically what we should be able to do is to just run the job again (i.e. the job should be idempotent). Can you find a way to make it so?

vojtechjelinek replied:
@seanlip This should be fixed now. The job is idempotent now.

seanlip replied:
Add individual tests that simulate the different cases for each of the possible user ID migration policies (for dependent models). This is crucial functionality and it is important we get it right and be prepared for any unexpected errors.

E.g. for _copy_model_with_new_id_and_user_id, have tests that:

- Setup a simulation where there are two models of this kind, one has been copied, the other is not (because of an earlier failure). Rerun the job. Check that it doesn't error (even though the first model was deleted after the move), it only migrates one model, etc.
- Setup a simulation where there are two models of this kind, and neither have been copied. Rerun the job. Check that it doesn't error and does the correct thing.
- Setup a simulation where there are two models of this kind, and both have already been copied. Rerun the job. Check that it has no effect on the models and does nothing.

Do similarly comprehensive unit tests for each case. Think through the different breakage possibilities and write one test for each of them. That will give more assurance that we are covered for the different eventualities and that the job is truly idempotent.

seanlip reviewed (changes requested):
Thanks @vojtechjelinek! Left some comments.

vojtechjelinek committed:
Address comments

vojtechjelinek commented:
@seanlip @kevinlee12 PTAL

kevinlee12 reviewed (approved):
lgtm for code owner files

seanlip reviewed (changes requested):
Thanks -- just one follow-up question.

vojtechjelinek committed:
Merge branch 'wipeout-all-users' of https://github.com/vojtechjelinek/oppia into develop

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

vojtechjelinek committed:
Make migration job idempotent

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

vojtechjelinek committed:
Replace the UserSettingsModel first

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

vojtechjelinek committed:
Fix tests

kevinlee12 commented:
@vojtechjelinek is it possible for this job to not run builds on Travis?

kevinlee12 commented:
If so, please remove it from the Travis queue, thanks!

vojtechjelinek commented:
@seanlip PTAL

seanlip reviewed (changes requested):
Thanks @vojtechjelinek! It looks better now, but I left a comment because I think this needs more careful verification (especially since this was missed before).

vojtechjelinek committed:
Add more tests

vojtechjelinek commented:
@seanlip I have added the tests, PTAL.

seanlip reviewed (commented):
Looks pretty good to me. Thanks!

Please send me instructions and I'd be happy to do a trial run.

vojtechjelinek commented:
@seanlip Can we merge this? This PR is already quite big and merging it wouldn't mean that it had to be run.

I will also deploy it to the backup server and send you the instructions.

kevinlee12 commented:
@vojtechjelinek please merge from develop as it doesn't have the latest CircleCI config.

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

vojtechjelinek committed:
Reduce the number of models in a transaction

vojtechjelinek committed:
Migrate contributors_summary in explorations and collections

vojtechjelinek committed:
Migrate UserSettingsModel separately

vojtechjelinek committed:
Fix linting

vojtechjelinek commented:
@seanlip PTAL (last 3 commit are relevant)

I have separated the creation of `UserSettingsModel` into separate job (after some thinking this came as the most efficient thing to do).

Also, I have added code to replace the user IDs in the `contributors_summary` in `ExpSummaryModel` and `CollectionSummaryModel`.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -47,11 +47,57 @@ class MissingUserException(Exception):
-    models that use the user id. This migration doesn't handle snapshot content
-    models that can contain user ID, these are handled by
-    SnapshotsUserIdMigrationJob.
+    """One-off job for replacing the old user ids wit new user ids in all
```
wit --> with

vojtechjelinek replied:
Done

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -47,11 +47,57 @@ class MissingUserException(Exception):
+class CreateNewUsersMigrationJob(jobs.BaseMapReduceOneOffJobManager):
+    """One-off job for creating new UserSettingsModels with new user ids set.
+    This migration doesn't handle the replacement of old user ids in the other
+    models, this is done by the UserIdMigrationJob.
```
Here and in other jobs: if two jobs need to be run one after the other to ensure consistent results, add this information in the docstrings of the jobs (to be clear that this job isn't truly independent).

vojtechjelinek replied:
Done

seanlip reviewed (commented):
Thanks! Left a few notes.

vojtechjelinek committed:
Address comments

vojtechjelinek committed:
Fix coverage

vojtechjelinek committed:
Add CreateNewUsersMigrationJob to registry

vojtechjelinek committed:
Always check that contributors_summary contains user ID

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

# Conflicts:
#	core/domain/activity_jobs_one_off.py

kevinlee12 commented:
@vojtechjelinek is this waiting for Sean's reply?

seanlip commented:
Actually, I'm confused too. @vojtechjelinek -- is this PR meant for me to look at, or is it waiting for the other one to go in (the one which regenerates contributor summaries)?

vojtechjelinek commented:
We are waiting for #8961

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

# Conflicts:
#	core/domain/exp_jobs_one_off.py
#	core/domain/exp_jobs_one_off_test.py

vojtechjelinek committed:
Add logging

vojtechjelinek committed:
Merge branch 'wipeout-all-users' of https://github.com/vojtechjelinek/oppia into wipeout-all-users

vojtechjelinek committed:
Fix migration

vojtechjelinek committed:
Rework migration for models with one user ID field

vojtechjelinek committed:
Add logging for _copy_model_with_new_id_and_user_id

vojtechjelinek committed:
Fix user ID migration for user models

vojtechjelinek committed:
Modify the shards count for the migration jobs

vojtechjelinek committed:
Fix UserSettingsModel verify

vojtechjelinek committed:
Add verify model

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

vojtechjelinek commented:
@seanlip PTAL, changes I've made:

- added `creator_ids` from `UserSubscriptionsModel` to the migration
- added `subscriber_ids` from `UserSubscribersModel` to the migration
- edited the shard count for all the jobs
- modified `verify_model_user_ids_exist` in `UserSettingsModel` since this model is always migrated

Btw, I went again through all the models to check if we didn't miss any other user IDs and didn't find anything else.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -243,7 +242,9 @@ def map(user_model):
                   base_models.USER_ID_MIGRATION_POLICY.CUSTOM):
-                model_class.migrate_model(old_user_id, new_user_id)
+                output = model_class.migrate_model(old_user_id, new_user_id)
+            if output:
```
if output is not None

vojtechjelinek replied:
Done

seanlip reviewed (commented):
Thanks, I read all the changes and they seem reasonable. Just noticed one nit.

vojtechjelinek committed:
Fix backend tests

vojtechjelinek committed:
Address comments

vojtechjelinek committed:
Fix coverage

seanlip commented on core/storage/audit/gae_models_test.py:
```diff
@@ -69,14 +69,6 @@ def test_get_user_id_migration_policy(self):
             audit_models.RoleQueryAuditModel.get_user_id_migration_policy(),
             base_models.USER_ID_MIGRATION_POLICY.COPY_AND_UPDATE_ONE_FIELD)
 
-    def test_get_user_id_migration_field(self):
```
Please justify the deletion of this test.

vojtechjelinek replied:
This model no longer has `get_user_id_migration_field` method since it is in different migration category.

vojtechjelinek committed:
Generate ID with prefix

vojtechjelinek committed:
Fix output truncation for SUCCESS cases

vojtechjelinek committed:
Migrate last_nonempty_message_author_id

vojtechjelinek committed:
Generate user id

vojtechjelinek committed:
Fix verify_model_user_ids_exist in feedback

vojtechjelinek committed:
Add tests for feedback

vojtechjelinek committed:
Fix tests for user id migration

vojtechjelinek committed:
Merge branch 'develop' of github.com:oppia/oppia into wipeout-all-users

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -88,10 +88,7 @@ def _replace_model():
-            yield (key, len(old_new_user_id_tuples))
-        else:
-            yield (key, old_new_user_id_tuples)
+        yield (key, len(old_new_user_id_tuples))
```
One note -- if you need any of these for debugging, one thing you could also do is to print the first ten items and then the number of additional items after that.

vojtechjelinek replied:
Not needed. Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -245,7 +242,7 @@ def map(user_model):
     def reduce(key, old_new_user_id_tuples):
         """Implements the reduce function for this job."""
-        yield (key, old_new_user_id_tuples)
+        yield (key, len(old_new_user_id_tuples))
```
Ditto.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -427,8 +424,9 @@ def entity_classes_to_map_over(cls):
         gae_id = user_model.gae_id
-        if (len(user_model.id) != user_models.USER_ID_LENGTH or
-                not all(c.islower() for c in user_model.id)):
+        if (not user_model.id.startswith('uid_') or
```
I suggest you write a standalone function that checks whether a user ID is in the correct "new" format, and write unit tests for that function as well. Then call that function here.

vojtechjelinek replied:
Done.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -220,7 +222,7 @@ def get_new_id(cls, unused_entity_name):
-            new_id = ''.join(
+            new_id = 'uid_%s' % ''.join(
                 random.choice(string.ascii_lowercase)
                 for _ in python_utils.RANGE(USER_ID_LENGTH))
```
Should this be USER_ID_RANDOM_PART_LENGTH?

You'll probably also want a backend unit test to ensure that when a new user ID is generated, its length matches USER_ID_LENGTH.

vojtechjelinek replied:
Done.

seanlip reviewed (changes requested):
@vojtechjelinek FYI, took a look and left 4 comments.

vojtechjelinek committed:
Separate user ID validation into method

vojtechjelinek committed:
Fix prod validation job

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Add deletion job

vojtechjelinek committed:
Fix tests

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -897,3 +897,35 @@ def map(rights_snapshot_model):
+
+
+class DeleteAllUserIdsVerificationJob(jobs.BaseMapReduceOneOffJobManager):
+    """Delete all RightsAllUsersModels from the datastore This needs to be done
```
Note: Missing period after "datastore".

vojtechjelinek replied:
Done.

kevintab95 reviewed (approved):
LGTM for code-owner file!

vojtechjelinek committed:
Address comments

vojtechjelinek committed:
Fix test coverage

vojtechjelinek committed:
Remove GeneralFeedbackThreadUserOneOffJob and require fields in GeneralFeedbackThreadUserModel

vojtechjelinek committed:
Reformat docstrings

vojtechjelinek committed:
Add migration job for snapshot metadata

vojtechjelinek committed:
Fix backend tests

vojtechjelinek committed:
Add verification job and tests

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

vojtechjelinek committed:
Add tests

vojtechjelinek commented:
@seanlip PTAl, added a code to migrate fields in `commit_cmds`

seanlip commented:
@vojtechjelinek Could you please make sure to reply to all the existing comments before requesting a review? Thanks!

seanlip commented:
(Also, in future, please update the assignees list accordingly if the ball is in my court.)

vojtechjelinek commented:
@seanlip Sorry, done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -42,6 +44,65 @@
 
 
+def replace_gae_ids(gae_ids):
+    """Replace GAE IDs with user IDs in list.
```
This is unclear -- what list are you referring to? I can't understand the docstring.

seanlip replied:
Also the implementation code in this function doesn't seem to be "replacing" anything?

vojtechjelinek replied:
Done, changed the name of the function.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -42,6 +44,65 @@
+        gae_ids: list(str). GAE IDs which should be replaced.
+
+    Returns:
+        list(str). New user IDs.
```
Also unclear, please state with more detail what these actually represent.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -42,6 +44,65 @@
+        list(str). New user IDs.
+
+    Raises:
+        MissingUserException: UserSettingsModel with GAE ID doesn't exist.
```
No UserSettingsModel exists for at least one of the given GAE IDs.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -42,6 +44,65 @@
+            new_ids.append(gae_id)
+        else:
+            user_settings_model = (
+                user_models.UserSettingsModel.get_by_gae_id(gae_id))
```
Do get_multi() here

vojtechjelinek replied:
Since the previous runs of migration didn't use `get_multi` here and there were quite quick anyway, and that the implementation of `get_multi` for the GAE ID will be a bit complicated, I would argue that we do not need it here.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -42,6 +44,65 @@
+    return new_ids
+
+
+def replace_gae_id(gae_id):
```
Same note -- "replace" suggests something is being replaced but I do not see an edit action here. Do you mean get_user_id_corresponding_to_gae_id() or similar?

vojtechjelinek replied:
Done, renamed the function.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -42,6 +44,65 @@
+        str. New user ID.
+
+    Raises:
+        MissingUserException: UserSettingsModel with GAE ID doesn't exist.
```
Add "the given" before "GAE ID"

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -42,6 +44,65 @@
+
+
+def are_commit_cmds_role_change(commit_cmds):
+    """Check if commit_cmds are of role_change type.
```
This doesn't match the returns docstring, which includes the additional constraint of there being exactly 1 cmd.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -401,6 +428,155 @@ def reduce(key, ids):
+            transaction_services.run_in_transaction(_put_both_models)
+
+    @staticmethod
+    def _migrate_exploration(snapshot_model):
```
Too much copy-paste -- just have one function for this and the above, and a local variable storing  exp_models.ExplorationCommitLogEntryModel. or collection_models.CollectionCommitLogEntryModel as needed. 

(Or have two functions but have both of them delegate to the same function -- e.g. have _migrate_collection just call _migrate_rights_snapshot_metadata_model(collection_models.CollectionCommitLogEntryModel, snapshot_model) or similar.)

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -401,6 +428,155 @@ def reduce(key, ids):
+                    'rights-%s-%s' % (
+                        snapshot_model.get_unversioned_instance_id(),
+                        snapshot_model.get_version_string())))
+            commit_cmd = commit_log_model.commit_cmds[0]
```
Don't reuse vars, give them different names instead.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -401,6 +428,155 @@ def reduce(key, ids):
+                    'rights-%s-%s' % (
+                        snapshot_model.get_unversioned_instance_id(),
+                        snapshot_model.get_version_string())))
+            commit_cmd = commit_log_model.commit_cmds[0]
```
Ditto, don't reuse vars. When a dev reads code and sees two vars named the same, the automatic assumption is that they represent the same thing, but in this case they don't.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -539,11 +715,8 @@ def map(model):
-                yield ('SUCCESS_NONE - %s' % model_class.__name__, model.id)
-            elif (BaseModelsUserIdsHaveUserSettingsVerificationJob
-                  ._check_id_and_user_id_exist(model.id, user_id)):
+            if (BaseModelsUserIdsHaveUserSettingsVerificationJob
```
Any reason to drop the previous "if user_id is None" clause? Still seems good to have it for safety, right?

vojtechjelinek replied:
It can't any longer happen that the `user_id` would be `None` in this branch of code, all the models with `COPY_AND_UPDATE_ONE_FIELD` have the field containing user ID always set to something. (This was verified in the previous runs of the migration)

seanlip replied:
You might still consider keeping it for safety. If someone adds a new model type before the actual migration and there's an error in the field, this would at least catch it.

vojtechjelinek replied:
Done, readded.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -899,9 +1073,106 @@ def reduce(key, ids):
+        if are_commit_cmds_role_change(snapshot_model.commit_cmds):
+            all_user_ids = set(all_users_model.all_user_ids)
+            all_user_ids.add(snapshot_model.commit_cmds[0]['assignee_id'])
+            all_users_model.all_user_ids = list(all_user_ids)
```
Do you care about order (do you need "sorted")? Might be worth sorting anyway, as a general best practice, so that you have a canonical "correct" representation.

vojtechjelinek replied:
I would argue that we shouldn't impose any order on the fields that contain multiple user IDs. Even tho the fields are list I think we should think about them as sets and these are usually not sorted. The ordering is not used anywhere else and I don't see any benefit in adding it here.

We can maybe discuss this in some call but I think if we want to do this anyway it would make sense to do it in a different PR.

vojtechjelinek replied:
Also, ordering of the IDs can prolong the jobs times by a few seconds without any clear benefit.

vojtechjelinek replied:
See https://github.com/oppia/oppia/pull/8505#discussion_r426188943

seanlip replied:
I'm not too bothered about the time complexity because the lists are very small and a few seconds is not a big deal.

But it looks like these are long-lived fields, right? If so, then let's please keep them sorted. It is better to have more constraints on the models so that, when reading from them, you can assume more things. (If the models in question were temporary I wouldn't care as much.)

vojtechjelinek replied:
Nope, this model will only be used for the migration and remove afterward.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -899,9 +1073,106 @@ def reduce(key, ids):
+            elif commit_cmds[0]['cmd'] == topic_domain.CMD_REMOVE_MANAGER_ROLE:
+                all_user_ids.add(
+                    snapshot_model.commit_cmds[0]['removed_user_id'])
+            all_users_model.all_user_ids = list(all_user_ids)
```
Ditto.

seanlip replied:
This still applies -- see above.

seanlip commented on core/domain/user_id_migration_test.py:
```diff
@@ -2141,6 +2579,247 @@ def test_multiple_rights(self):
+            exp_models.ExplorationRightsAllUsersModel.get_by_id(self.EXP_1_ID)
+            .all_user_ids)
+
+    def test_one_topic_rights_cahnge_role(self):
```
"change" is misspelled.

vojtechjelinek replied:
Done.

seanlip commented on core/storage/classifier/gae_models.py:
```diff
@@ -245,7 +245,7 @@ def get_user_id_migration_policy():
     def _generate_id(cls, exp_id, exp_version, state_name):
         """Generates a unique ID for the Classifier Exploration Mapping of the
-        form {{exp_id}}.{{exp_version}}.{{utf8_encoded_state_name}}
+        form [exp_id].[exp_version].[state_name].
```
I see these sorts of fixes every so often. Maybe just do a pass through the whole storage layer and ensure each of them has a clear ID specced in the correct format?

seanlip replied:
Ah -- actually, looks like you did this already :) Thanks!

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -504,8 +504,8 @@ class GeneralFeedbackThreadUserModel(base_models.BaseModel):
 
-    user_id = ndb.StringProperty(required=False, indexed=True)
-    thread_id = ndb.StringProperty(required=False, indexed=True)
+    user_id = ndb.StringProperty(required=True, indexed=True)
```
This could break existing models, right? Feedback threads can be started anonymously.

vojtechjelinek replied:
No, this was fixed in #8481. Anonymous models don't have `GeneralFeedbackThreadUserModel`.

seanlip reviewed (changes requested):
Thanks @vojtechjelinek! Took a pass.

srijanreddy98 commented:
Ping @vojtechjelinek

vojtechjelinek committed:
Address comments

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -70,23 +71,24 @@ def replace_gae_ids(gae_ids):
 
 def are_commit_cmds_role_change(commit_cmds):
-    """Check if commit_cmds are of role_change type.
+    """Check if commit_cmds are of a role change type and there just one commit.
```
there --> there is

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -435,55 +446,28 @@ class SnapshotsMetadataUserIdMigrationJob(jobs.BaseMapReduceOneOffJobManager):
 
     @staticmethod
-    def _migrate_collection(snapshot_model):
+    def _migrate_model(snapshot_model, commit_model_type):
```
I think you mean commit_model_class

vojtechjelinek replied:
Done.

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -435,55 +446,28 @@ class SnapshotsMetadataUserIdMigrationJob(jobs.BaseMapReduceOneOffJobManager):
         Args:
             snapshot_model: CollectionRightsSnapshotMetadataModel. The model
                 that contains the old user IDs.
+            commit_model_type: ndb.Model. The type of the commit log model.
```
ndb.Model is the type of an instance of the model, but you're passing in the class, which would have a different type, right?

vojtechjelinek replied:
Yep, done.

vojtechjelinek committed:
Address comments

seanlip reviewed (approved):
LGTM, no further comments. Thanks @vojtechjelinek!

seanlip commented:
Please note the failing backend tests though.

seanlip commented:
(Also next time please remember to reassign the reviewer when you want them to look at the PR!)

vojtechjelinek committed:
Fix coverage

kevinlee12 commented:
Looks like all checks are green, is this ready for merge?

vojtechjelinek commented:
@kevinlee12 Nope, this needs to be tested on the backupcould you please mark it as 'Requested changes'.

kevinlee12 reviewed (dismissed):
Locking as requested

kevinlee12 commented:
Sure, let me know when to merge.

vojtechjelinek committed:
Merge branch 'develop' into wipeout-all-users

nishantwrp commented:
@vojtechjelinek Please resolve the merge conflicts!

vojtechjelinek committed:
Fix SnapshotsMetadataUserIdMigrationJob

seanlip commented on core/domain/user_id_migration.py:
```diff
@@ -465,6 +469,12 @@ def _migrate_model(snapshot_model, commit_model_class):
+                snapshot_model.put(update_last_updated_time=False)
+                snapshot_class_name = snapshot_model.__class__.__name__
+                return (
+                    'SUCCESS_MISSING_COMMIT - %s' % snapshot_class_name,
```
Is a missing commit actually a success? Should we be logging those cases?

vojtechjelinek replied:
Success in the sense that we can fix this after the migration.

vojtechjelinek committed:
Make jobs idempotent

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Mark job results clearly as success

vojtechjelinek committed:
Merge branch 'develop' of github.com:oppia/oppia into wipeout-all-users

vojtechjelinek committed:
Fix tests

seanlip commented:
Confirming that this PR passes on the backup server and can be merged!

vojtechjelinek commented:
@kevinlee12 Please approve this it is already tested on the backup server and needs to be merged for this release.
